# Анализ базы данных.

<h1>Оглавление:<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Описание-проекта:" data-toc-modified-id="Описание-проекта:-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span>Описание проекта:</a></span></li><li><span><a href="#Описание-данных:" data-toc-modified-id="Описание-данных:-0.2"><span class="toc-item-num">0.2&nbsp;&nbsp;</span>Описание данных:</a></span></li><li><span><a href="#Схема-данных:" data-toc-modified-id="Схема-данных:-0.3"><span class="toc-item-num">0.3&nbsp;&nbsp;</span>Схема данных:</a></span></li><li><span><a href="#Задания:" data-toc-modified-id="Задания:-0.4"><span class="toc-item-num">0.4&nbsp;&nbsp;</span>Задания:</a></span></li></ul></li><li><span><a href="#Импортирую-необходимые-библиотеки:" data-toc-modified-id="Импортирую-необходимые-библиотеки:-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Импортирую необходимые библиотеки:</a></span></li><li><span><a href="#Устанавливаю-параметры:" data-toc-modified-id="Устанавливаю-параметры:-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Устанавливаю параметры:</a></span></li><li><span><a href="#Изучаю-общую-информацию." data-toc-modified-id="Изучаю-общую-информацию.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Изучаю общую информацию.</a></span><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Провожу-исследовательский-анализ-данных-(EDA)." data-toc-modified-id="Провожу-исследовательский-анализ-данных-(EDA).-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Провожу исследовательский анализ данных (EDA).</a></span><ul class="toc-item"><li><span><a href="#Считаю,-сколько-книг-вышло-после-1-января-2000-года." data-toc-modified-id="Считаю,-сколько-книг-вышло-после-1-января-2000-года.-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Считаю, сколько книг вышло после 1 января 2000 года.</a></span></li><li><span><a href="#Для-каждой-книги-считаю-количество-обзоров-и-среднюю-оценку." data-toc-modified-id="Для-каждой-книги-считаю-количество-обзоров-и-среднюю-оценку.-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Для каждой книги считаю количество обзоров и среднюю оценку.</a></span></li><li><span><a href="#Определяю-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц." data-toc-modified-id="Определяю-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц.-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Определяю издательство, которое выпустило наибольшее число книг толще 50 страниц.</a></span></li><li><span><a href="#Определяю-автора-с-самой-высокой-средней-оценкой-книг---учитывая-только-книги-с-50-и-более-оценками." data-toc-modified-id="Определяю-автора-с-самой-высокой-средней-оценкой-книг---учитывая-только-книги-с-50-и-более-оценками.-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Определяю автора с самой высокой средней оценкой книг - учитывая только книги с 50 и более оценками.</a></span></li><li><span><a href="#Считаю-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок." data-toc-modified-id="Считаю-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок.-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Считаю среднее количество обзоров от пользователей, которые поставили больше 50 оценок.</a></span></li></ul></li><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></div>

### Описание проекта:
Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы - и бросились создавать приложения для тех, кто любит читать.
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Задача - проанализировать базу данных. В ней - информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.
### Описание данных:
Таблица `books`, содержит данные о книгах:  
- **`book_id`** - идентификатор книги;
- **`author_id`** - идентификатор автора;
- **`title`** - название книги;
- **`num_pages`** - количество страниц;
- **`publication_date`** - дата публикации книги;
- **`publisher_id`** - идентификатор издателя.

Таблица `authors`, содержит данные об авторах:   
- **`author_id`** - идентификатор автора;
- **`author`** - имя автора.

Таблица `publishers`, содержит данные об издательствах:
- **`publisher_id`** - идентификатор издательства;
- **`publisher`** - название издательства.

Таблица `ratings`, содержит данные о пользовательских оценках книг:
- **`rating_id`** - идентификатор оценки;
- **`book_id`** - идентификатор книги;
- **`username`** - имя пользователя, оставившего оценку;
- **`rating`** - оценка книги.

Таблица `reviews`, содержит данные о пользовательских обзорах:
- **`review_id`** - идентификатор обзора;
- **`book_id`** - идентификатор книги;
- **`username`** - имя автора обзора;
- **`text`** - текст обзора.

### Схема данных:
![jupyter](https://pictures.s3.yandex.net/resources/scheme_1589269096.png)

### Задания:
- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц - так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг - учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Импортирую необходимые библиотеки:

In [1]:
import pandas as pd
from sqlalchemy import create_engine

## Устанавливаю параметры:

In [2]:
db_config = {'user':'praktikum_student', # имя пользователя
              'pwd':'Sdf4$2;d-d30pp', # пароль
             'host':'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port':6432, # порт подключения
               'db':'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])

engine = create_engine(connection_string, connect_args = {'sslmode':'require'}) # сохраняю коннектор

## Изучаю общую информацию.

In [3]:
def table_info(table):
    display(pd.io.sql.read_sql('''SELECT * FROM {} LIMIT 5;'''.format(table), con = engine))
    print('-' * 50)
    pd.io.sql.read_sql('''SELECT * FROM {};'''.format(table), con = engine).info()
    print('-' * 50)
    display(pd.io.sql.read_sql('''SELECT * FROM {};'''.format(table), con = engine).describe(include = 'all'))
    print('-' * 50)
    print('Количество явных дубликатов:', pd.io.sql.read_sql('''SELECT * FROM {};'''.format(table), con = engine).duplicated().sum())

In [4]:
table_info('books')

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
--------------------------------------------------


book_id    author_id                title   num_pages  \
count   1000.000000  1000.000000                 1000  1000.00000   
unique          NaN          NaN                  999         NaN   
top             NaN          NaN  Memoirs of a Geisha         NaN   
freq            NaN          NaN                    2         NaN   
mean     500.500000   320.417000                  NaN   389.11100   
std      288.819436   181.620172                  NaN   229.39014   
min        1.000000     1.000000                  NaN    14.00000   
25%      250.750000   162.750000                  NaN   249.00000   
50%      500.500000   316.500000                  NaN   352.00000   
75%      750.250000   481.000000                  NaN   453.00000   
max     1000.000000   636.000000                  NaN  2690.00000   

       publication_date  publisher_id  
count              1000   1000.000000  
unique              618           NaN  
top          2006-08-29           NaN  
freq                 10           NaN  
mean                NaN    171.270000  
std                 NaN     99.082685  
min                 NaN      1.000000  
25%                 NaN     83.000000  
50%                 NaN    177.500000  
75%                 NaN    258.000000  
max                 NaN    340.000000

--------------------------------------------------
Количество явных дубликатов: 0


In [5]:
pd.io.sql.read_sql(''' SELECT *
                         FROM books
                        WHERE title = 'Memoirs of a Geisha';
                   ''', con = engine)

book_id  author_id                title  num_pages publication_date  \
0      426         39  Memoirs of a Geisha        434       2005-11-15   
1      427         39  Memoirs of a Geisha        503       2005-11-22   

   publisher_id  
0           241  
1           311

Таблица `books` содержит **1000** записей по **6** столбцам. Типы данных соответствуют содержимому столбцов. Пропусков и явных дубликатов нет. Данные изначально очень хорошего качества. 

Одна книга - **Memoirs of a Geisha**, представлена дважды, под разными издательствами, остальные **999** книг уникальны. Количество страниц от **14** до **2690**, в среднем - **352**. Представлено **636** уникальных авторов и **340** уникальных издательств. Самая частая дата публикации книги - **29 августа 2006**, в этот день выпущено **10** книг.

In [6]:
table_info('authors')

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
--------------------------------------------------


author_id      author
count   636.000000         636
unique         NaN         636
top            NaN  A.S. Byatt
freq           NaN           1
mean    318.500000         NaN
std     183.741666         NaN
min       1.000000         NaN
25%     159.750000         NaN
50%     318.500000         NaN
75%     477.250000         NaN
max     636.000000         NaN

--------------------------------------------------
Количество явных дубликатов: 0


Таблица `authors` содержит **636** записей по **2** столбцам. Типы данных соответствуют содержимому столбцов. Пропусков и явных дубликатов нет. Данные изначально очень хорошего качества. 

Это **636** уникальных авторов и их id.

In [7]:
table_info('publishers')

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
--------------------------------------------------


publisher_id publisher
count     340.000000       340
unique           NaN       340
top              NaN       Ace
freq             NaN         1
mean      170.500000       NaN
std        98.293777       NaN
min         1.000000       NaN
25%        85.750000       NaN
50%       170.500000       NaN
75%       255.250000       NaN
max       340.000000       NaN

--------------------------------------------------
Количество явных дубликатов: 0


Таблица `publishers` содержит **340** записей по **2** столбцам. Типы данных соответствуют содержимому столбцов. Пропусков и явных дубликатов нет. Данные изначально очень хорошего качества. 

Это **340** уникальных издательств и их id.

In [8]:
table_info('ratings')

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
--------------------------------------------------


rating_id      book_id    username       rating
count   6456.000000  6456.000000        6456  6456.000000
unique          NaN          NaN         160          NaN
top             NaN          NaN  martinadam          NaN
freq            NaN          NaN          56          NaN
mean    3228.500000   510.574195         NaN     3.928284
std     1863.831001   284.141636         NaN     0.943303
min        1.000000     1.000000         NaN     1.000000
25%     1614.750000   291.000000         NaN     3.000000
50%     3228.500000   506.000000         NaN     4.000000
75%     4842.250000   750.000000         NaN     5.000000
max     6456.000000  1000.000000         NaN     5.000000

--------------------------------------------------
Количество явных дубликатов: 0


Таблица `ratings` содержит **6456** записей по **4** столбцам. Типы данных соответствуют содержимому столбцов. Пропусков и явных дубликатов нет. Данные изначально очень хорошего качества. 

Представлена информация о **160** уникальных пользователях, которые оставили **6456** оценок для **1000** книг. Средняя оценка - **4**. Самый активный пользователь - **martinadam**, он оценил **56** книг.

In [9]:
table_info('reviews')

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
--------------------------------------------------


review_id      book_id username  \
count   2793.000000  2793.000000     2793   
unique          NaN          NaN      160   
top             NaN          NaN  susan85   
freq            NaN          NaN       29   
mean    1397.000000   504.693161      NaN   
std      806.413976   288.472931      NaN   
min        1.000000     1.000000      NaN   
25%      699.000000   259.000000      NaN   
50%     1397.000000   505.000000      NaN   
75%     2095.000000   753.000000      NaN   
max     2793.000000  1000.000000      NaN   

                                                     text  
count                                                2793  
unique                                               2793  
top     Mention society tell send professor analysis. ...  
freq                                                    1  
mean                                                  NaN  
std                                                   NaN  
min                                                   NaN  
25%                                                   NaN  
50%                                                   NaN  
75%                                                   NaN  
max                                                   NaN

--------------------------------------------------
Количество явных дубликатов: 0


Таблица `reviews` содержит **2793** записи по **4** столбцам. Типы данных соответствуют содержимому столбцов. Пропусков и явных дубликатов нет. Данные изначально очень хорошего качества. 

Представлена информация о **160** уникальных пользователях, которые оставили **2793** обзора для **1000** книг. Самый активный обозреватель - **susan85**, он оставил **29** обзоров.

### Вывод:
В пяти имеющихся таблицах пропущенных значений и явных дубликатов нет. Типы данных соответствуют содержимому столбцов. Данные изначально очень хорошего качества. Можно приступать к следующему шагу.

## Провожу исследовательский анализ данных (EDA).

In [10]:
def request(query):
    display(pd.io.sql.read_sql(query, con = engine))

### Считаю, сколько книг вышло после 1 января 2000 года.

In [11]:
query = ''' SELECT COUNT(book_id) count_book
              FROM books
             WHERE publication_date >= DATE('2000-01-01');
        '''
request(query)

count_book
0         821

После 1 января 2000 года вышла **821** книга.

### Для каждой книги считаю количество обзоров и среднюю оценку.

In [12]:
query = ''' WITH 
            request_1 AS (
                SELECT book_id, COUNT(review_id) count_review 
                  FROM reviews
              GROUP BY book_id
            ),
            request_2 AS (
                SELECT book_id, AVG(rating) average_raiting
                  FROM ratings 
              GROUP BY book_id
            )
            SELECT title, SUM(count_review) count_review, ROUND(AVG(average_raiting), 2) average_raiting
              FROM books 
         LEFT JOIN request_1
                ON books.book_id = request_1.book_id
         LEFT JOIN request_2 
                ON books.book_id = request_2.book_id
          GROUP BY books.title
          ORDER BY average_raiting DESC;
        '''
request(query)

title  count_review  \
0             Evil Under the Sun (Hercule Poirot  #24)           2.0   
1    The Demon-Haunted World: Science as a Candle i...           2.0   
2    The War of Art: Break Through the Blocks & Win...           2.0   
3                                   The Woman in Black           2.0   
4                        Emily of New Moon (Emily  #1)           2.0   
..                                                 ...           ...   
994  The World Is Flat: A Brief History of the Twen...           3.0   
995                  His Excellency: George Washington           2.0   
996                                      Drowning Ruth           3.0   
997                                              Junky           2.0   
998                               Harvesting the Heart           2.0   

     average_raiting  
0               5.00  
1               5.00  
2               5.00  
3               5.00  
4               5.00  
..               ...  
994             2.25  
995             2.00  
996             2.00  
997             2.00  
998             1.50  

[999 rows x 3 columns]

Готово.

### Определяю издательство, которое выпустило наибольшее число книг толще 50 страниц.

In [13]:
query = ''' SELECT publisher, COUNT(book_id) count_book 
              FROM books 
         LEFT JOIN publishers 
                ON books.publisher_id = publishers.publisher_id
             WHERE num_pages > 50
          GROUP BY publisher
          ORDER BY count_book DESC
             LIMIT 1;
        '''
request(query)

publisher  count_book
0  Penguin Books          42

Наибольшее число книг толще 50 страниц выпустило издательство **Penguin Books**.

Количество книг - **42**.

### Определяю автора с самой высокой средней оценкой книг - учитывая только книги с 50 и более оценками.

In [14]:
query = ''' WITH 
            request AS (
                SELECT author_id, title, AVG(rating) average_raiting
                  FROM books 
             LEFT JOIN ratings 
                    ON books.book_id = ratings.book_id 
              GROUP BY title, author_id
                HAVING COUNT(rating_id) >= 50
            )
            SELECT author, ROUND(AVG(average_raiting), 2) average_raiting
              FROM request
         LEFT JOIN authors 
                ON request.author_id = authors.author_id
          GROUP BY author
          ORDER BY average_raiting DESC
             LIMIT 1;
        '''
request(query)

author  average_raiting
0  J.K. Rowling/Mary GrandPré             4.28

Автор с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками - **J.K. Rowling/Mary GrandPré**.

Средняя оценка - **4.28**.

### Считаю среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [15]:
query = ''' WITH
            request_1 AS (
                SELECT username 
                  FROM ratings
              GROUP BY username
                HAVING COUNT(rating_id) > 50
            ),
            request_2 AS (
                SELECT COUNT(review_id) count_reviews
                  FROM request_1
             LEFT JOIN reviews 
                    ON request_1.username = reviews.username
              GROUP BY request_1.username
            ) 
            SELECT ROUND(AVG(count_reviews)) average_reviews
              FROM request_2;
        '''
request(query)

average_reviews
0             24.0

В среднем количество обзоров от пользователей, которые поставили больше 50 оценок - **24**.

## Вывод:
Я подключился к базе данных PostgreSQL и изучил основную информацию таблиц: `books`, `authors`, `publishers`, `ratings`, `reviews`:

В пяти имеющихся таблицах пропущенных значений и явных дубликатов нет. Типы данных соответствуют содержимому столбцов. Данные изначально очень хорошего качества. 
***
Исследовательский анализ данных показал, что:
- После 1 января 2000 года вышла **821** книга;
- Наибольшее число книг толще 50 страниц выпустило издательство **Penguin Books**. Количество книг - **42**;
- Автор с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками - **J.K. Rowling/Mary GrandPré**. Средняя оценка - **4.28**;
- В среднем количество обзоров от пользователей, которые поставили больше 50 оценок - **24**.

А также установил для каждой книги количество обзоров и среднюю оценку.